### PR TITLE
Optimization: allocate immutable Expression objects statically;

### PR DESCRIPTION
**Pull request by SergeiPavlov**
* Move Fun<,> to WellKnownTypes.FuncOfTArgTResultType

### CODE REVIEW

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 1]
```diff
@@ -33,19 +33,29 @@ namespace Xtensive.Orm.Linq
 {
   internal sealed partial class Translator
   {
+    private static readonly ParameterExpression parameterContextParam = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "context");
+    private static readonly ConstantExpression
+      nullKeyExpression = Expression.Constant(null, WellKnownOrmTypes.Key),
+      falseBoolExpression = Expression.Constant(false, WellKnownTypes.Bool),
+      trueBoolExpression = Expression.Constant(true, WellKnownTypes.Bool),
+      falseExpression = Expression.Constant(false),
+      trueExpression = Expression.Constant(true);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:42]: Why expression for the same constants are declared twice?

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 2]
```diff
@@ -646,6 +654,9 @@ internal static bool FilterBindings(MemberInfo mi, string name, Type type)
       return bindings;
     }
 
+    private static ConstantExpression SelectBoolConstantExpression(bool b) =>
+      b ? trueBoolExpression : falseBoolExpression;
+
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:659]: Place functions in the downward direction.

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs [hunk 3]
```diff
@@ -20,6 +21,12 @@ namespace Xtensive.Orm.Rse.Transformation
 {
   internal sealed class RedundantColumnRemover : ColumnMappingInspector
   {
+    private static readonly MethodInfo selectMethodInfo = WellKnownTypes.Enumerable
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs:24]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/Internals/ApplyFilterRewriter.cs [hunk 4]
```diff
@@ -17,15 +17,14 @@ namespace Xtensive.Orm.Rse.Transformation
 {
   internal sealed class ApplyFilterRewriter : ExpressionVisitor
   {
+    private static readonly ParameterExpression leftTupleParameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "leftTuple");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Rse/Transformation/Internals/ApplyFilterRewriter.cs:20]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs [hunk 5]
```diff
@@ -27,6 +27,8 @@ namespace Xtensive.Orm.Linq
   internal sealed partial class Translator : QueryableVisitor
   {
     private static readonly Type IEnumerableOfKeyType = typeof(IEnumerable<Key>);
+    private static readonly ParameterExpression tupleParameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "tuple");
+    private static readonly ParameterExpression parameterContextContextParameter = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "context");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs:31]: Naming policy violation. Correct name example is one line above.

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Materialization.cs [hunk 6]
```diff
@@ -23,6 +23,10 @@ namespace Xtensive.Orm.Linq
 {
   internal sealed partial class Translator
   {
+    private static readonly ParameterExpression parameterContext = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "parameterContext");
+    private static readonly ParameterExpression tupleReader = Expression.Parameter(typeof(RecordSetReader), "tupleReader");
+    private static readonly ParameterExpression session = Expression.Parameter(typeof(Session), "session");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Materialization.cs:28]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Linq/QueryHelper.cs [hunk 7]
```diff
@@ -33,10 +33,11 @@ public OwnerWrapper(TOwner owner)
       }
     }
 
+    private static readonly ParameterExpression tupleParameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "tuple");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/QueryHelper.cs:36]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Linq/ParameterAccessorFactory.cs [hunk 8]
```diff
@@ -18,6 +18,8 @@ internal static class ParameterAccessorFactory
     private static readonly MethodInfo GetParameterValueMethod =
       WellKnownOrmTypes.ParameterContext.GetMethod(nameof(ParameterContext.GetValue));
 
+    private static readonly ParameterExpression parameterContextArgument = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "context");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/ParameterAccessorFactory.cs:21]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs [hunk 9]
```diff
@@ -52,6 +52,9 @@ public void AddRange(IReadOnlyCollection<Type> newTypes)
       }
     }
 
+    private static readonly ParameterExpression paramContext = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "context");
+    private static readonly MethodInfo selectMethod = WellKnownMembers.Enumerable.Select.MakeGenericMethod(typeof(TItem), WellKnownOrmTypes.Tuple);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs:56]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs [hunk 10]
```diff
@@ -34,7 +34,7 @@ public MappingEntry(LambdaExpression calculatedColumn)
       }
     }
 
-    private readonly ParameterExpression calculatedColumnParameter;
+    private static readonly ParameterExpression calculatedColumnParameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "filteredRow");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs:37]: Naming policy violation.

File: Orm/Xtensive.Orm/Orm/EntitySetBase.cs [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// Copyright (C) 2008-2020 Xtensive LLC.
+// Copyright (C) 2008-2021 Xtensive LLC.
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/EntitySetBase.cs:1]: Why is this change here? No meaningful changes of the file

File: Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs [hunk 12]
```diff
@@ -25,18 +25,18 @@ namespace Xtensive.Orm.Building.Builders
 {
   internal class PartialIndexFilterBuilder : ExpressionVisitor
   {
+    private static readonly ParameterExpression parameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "tuple");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs:28]: Naming policy violation.

File: Orm/Xtensive.Orm/Linq/ConstantExtractor.cs [hunk 13]
```diff
@@ -21,9 +21,10 @@ namespace Xtensive.Linq
   /// </summary>
   public sealed class ConstantExtractor : ExpressionVisitor
   {
+    private static readonly ParameterExpression constantParameter = Expression.Parameter(WellKnownTypes.ObjectArray, "constants");
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Linq/ConstantExtractor.cs:24]: Naming policy violation.

**Review by alex-kulakov — CHANGES_REQUESTED** [review]

File: Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs [hunk 14]
```diff
@@ -31,9 +31,12 @@ internal class ExpressionMaterializer : PersistentExpressionVisitor
     private static readonly MethodInfo GetParameterValueMethod;
     private static readonly PropertyInfo ParameterContextProperty;
     private static readonly MethodInfo GetTupleParameterValueMethod;
+    private static readonly ParameterExpression tupleParameter = Expression.Parameter(WellKnownOrmTypes.Tuple, "tuple");
+
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs:35]: Probably merge effect. This commit  was result of some merges.
removed it.

File: Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs [hunk 15]
```diff
@@ -31,9 +31,12 @@ internal class ExpressionMaterializer : PersistentExpressionVisitor
...
+    private static readonly ParameterExpression materializationContextParameter = Expression.Parameter(WellKnownOrmTypes.ItemMaterializationContext, "mc");
+    private static readonly ConstantExpression typeReferenceAccuracyBaseTypeConstantExpression = Expression.Constant(TypeReferenceAccuracy.BaseType);
```
[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs:37]: renamed

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs (hunk 2 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs (hunk 10 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs:37]: what is naming policy here?
I see many lowercase private static names in the project

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs (hunk 1 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs]: Original code had usage of both forms.
Now I see they are semantically equal.

Removed duplication

File: Orm/Xtensive.Orm/Linq/ConstantExtractor.cs (hunk 13 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Linq/ConstantExtractor.cs]: fixed

File: Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs (hunk 12 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Building/Builders/PartialIndexFilterBuilder.cs]: fixed

File: Orm/Xtensive.Orm/Orm/EntitySetBase.cs (hunk 11 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/EntitySetBase.cs]: fixed

File: Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs (hunk 10 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Expressions/Visitors/IncludeFilterMappingGatherer.cs]: fixed

File: Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs (hunk 9 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs]: fixed

File: Orm/Xtensive.Orm/Orm/EntitySet{T}.cs [hunk 16]
```diff
@@ -62,6 +62,9 @@ public class EntitySet<TItem> : EntitySetBase,
     IQueryable<TItem>
     where TItem : IEntity
   {
+    private static readonly MemberExpression ownerPropertyExpression = Expression.Property(Expression.Constant(ownerParameter), ownerParameter.GetType()
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/EntitySet{T}.cs:65]:
```suggestion
    private static readonly MemberExpression OwnerPropertyExpression = Expression.Property(Expression.Constant(ownerParameter), ownerParameter.GetType()
```

File: Orm/Xtensive.Orm/Orm/EntitySet{T}.cs [hunk 17]
```diff
@@ -230,9 +233,7 @@ public Expression Expression
 
     private static IQueryable<TItem> GetItemsQuery(QueryEndpoint qe, FieldInfo field)
     {
-      var owner = Expression.Property(Expression.Constant(ownerParameter), ownerParameter.GetType()
-        .GetProperty("Value", WellKnownOrmTypes.Entity));
-      var queryExpression = QueryHelper.CreateEntitySetQuery(owner, field);
+      var queryExpression = QueryHelper.CreateEntitySetQuery(ownerPropertyExpression, field);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/EntitySet{T}.cs:236]:
```suggestion
      var queryExpression = QueryHelper.CreateEntitySetQuery(OwnerPropertyExpression, field);
```

File: Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs (hunk 9 above)
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs]: You renamed only one field of two.

File: Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs (hunk 15 above)
[the remark this answers is not in the document]
Comment (alex-kulakov) [reply Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs]: I wrote you exact name that follows the naming policy. What prevents you from using it? btw, you fixed two mistakes within these three lines. Additionally you have a lot of comments with "naming violation" for exact same reason, could you see a bit wider the place comment points?

[the remark this answers is not in the document]
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Materialization/ExpressionMaterializer.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs (hunk 9 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/ItemToTupleConverter{TItem}.cs]: done

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs (hunk 3 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Rse/Transformation/RedundantColumnRemover.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs (hunk 5 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Translator.Queryable.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Materialization.cs (hunk 6 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/Translator.Materialization.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/ParameterAccessorFactory.cs (hunk 8 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/ParameterAccessorFactory.cs]: done

File: Orm/Xtensive.Orm/Orm/Rse/Transformation/Internals/ApplyFilterRewriter.cs (hunk 4 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Rse/Transformation/Internals/ApplyFilterRewriter.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/QueryHelper.cs (hunk 7 above)
Comment (SergeiPavlov) [reply Orm/Xtensive.Orm/Orm/Linq/QueryHelper.cs]: done

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 18]
```diff
@@ -33,19 +33,27 @@ namespace Xtensive.Orm.Linq
 {
   internal sealed partial class Translator
   {
+    private static readonly ParameterExpression parameterContextParam = Expression.Parameter(WellKnownOrmTypes.ParameterContext, "context");
+    private static readonly ConstantExpression
+      nullKeyExpression = Expression.Constant(null, WellKnownOrmTypes.Key),
+      falseExpression = Expression.Constant(false),
+      trueExpression = Expression.Constant(true);
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:40]:
```suggestion
    private static readonly ConstantExpression
      NullKeyExpression = Expression.Constant(null, WellKnownOrmTypes.Key),
      FalseExpression = Expression.Constant(false),
      TrueExpression = Expression.Constant(true);
```

Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:40]:
```suggestion
    private static readonly ConstantExpression
      NullKeyExpression = Expression.Constant(null, WellKnownOrmTypes.Key),
      FalseExpression = Expression.Constant(false),
      TrueExpression = Expression.Constant(true);
```

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 19]
```diff
@@ -33,19 +33,27 @@ namespace Xtensive.Orm.Linq
...
+
     protected override Expression VisitTypeIs(TypeBinaryExpression tb)
     {
       var expression = tb.Expression;
       Type expressionType = expression.Type;
       Type operandType = tb.TypeOperand;
-      if (operandType.IsAssignableFrom(expressionType))
-        return Expression.Constant(true);
+      if (operandType.IsAssignableFrom(expressionType)) {
+        return trueExpression;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:48]:
```suggestion
        return TrueExpression;
```

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 20]
```diff
@@ -33,19 +33,27 @@ namespace Xtensive.Orm.Linq
...
+      }
 
       // Structure
       var memberType = expression.GetMemberType();
       if (memberType==MemberType.Structure
-        && WellKnownOrmTypes.Structure.IsAssignableFrom(operandType))
-        return Expression.Constant(false);
+        && WellKnownOrmTypes.Structure.IsAssignableFrom(operandType)) {
+        return falseExpression;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:55]:
```suggestion
        return FalseExpression;
```

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 21]
```diff
@@ -854,6 +861,9 @@ private Expression VisitBinaryRecursive(BinaryExpression binaryExpression, Binar
       return resultExpression;
     }
 
+    private static ConstantExpression SelectBoolConstantExpression(bool b) =>
+      b ? trueExpression : falseExpression;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:865]:
```suggestion
      b ? TrueExpression : FalseExpression;
```

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 22]
```diff
@@ -998,12 +1008,14 @@ private static IList<Expression> GetEntityFields(Expression expression, IEnumera
 
       Expression keyExpression;
 
-      if (expression.IsNull())
-        keyExpression = Expression.Constant(null, WellKnownOrmTypes.Key);
-      else if (IsConditionalOrWellknown(expression))
+      if (expression.IsNull()) {
+        keyExpression = nullKeyExpression;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:1012]:
```suggestion
        keyExpression = NullKeyExpression;
```

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 23]
```diff
@@ -1012,7 +1024,7 @@ private static IList<Expression> GetEntityFields(Expression expression, IEnumera
           expression = Expression.Convert(expression, WellKnownOrmInterfaces.Entity);
         keyExpression = Expression.Condition(
           isNullExpression,
-          Expression.Constant(null, WellKnownOrmTypes.Key),
+          nullKeyExpression,
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:1027]:
```suggestion
          NullKeyExpression,
```

**Review by alex-kulakov — CHANGES_REQUESTED** [review]
Almost ready, please, apply suggested changes and it will be ready to merge

File: Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs [hunk 24]
```diff
@@ -663,155 +669,156 @@ private Expression VisitBinaryRecursive(BinaryExpression binaryExpression, Binar
...
+            || left is ConstantExpression && ((ConstantExpression) left).Value == null)
+            return SelectBoolConstantExpression(binaryExpression.NodeType == ExpressionType.NotEqual);
+          if ((rightIsConstant && ExpressionEvaluator.Evaluate(right).Value == null)
+            || right is ConstantExpression && ((ConstantExpression) right).Value == null)
+            return SelectBoolConstantExpression(binaryExpression.NodeType == ExpressionType.NotEqual);
+          var leftEntitySetExpression = left as EntitySetExpression;
+          var rightEntitySetExpression = right as EntitySetExpression;
+          if (leftEntitySetExpression != null && rightEntitySetExpression != null) {
+            if (leftEntitySetExpression.Field != rightEntitySetExpression.Field) {
+              return falseExpression;
```
Comment (alex-kulakov) [inline Orm/Xtensive.Orm/Orm/Linq/Translator.Expressions.cs:699]:
```suggestion
              return FalseExpression;
```

**Review by alex-kulakov — APPROVED** [review]
